### ISSUE TITLE
Release timeline for 0.4.0.

### ISSUE THREAD

**Issue by alanmcruickshank** [sqlfluff/sqlfluff#470]
I'm putting in place some basic project process at the moment. As part of that I think we should have a place for the discussion around which features will and won't go into the 0.4.0 release.

The current project board is here: https://github.com/sqlfluff/sqlfluff/projects/2

As of 6th October, the shortlist looks pretty well populated. I think there are lots of excellent suggestions in there. Before bringing them all in I want to make sure that we have a clear path forward on each one.

The ones that I think more clarity before we can bring them in are:
- *The concept of a public API*.
  - https://github.com/sqlfluff/sqlfluff/issues/289
  - This one I'm happy to take hold of and get some clarity on, and I'm already working through some ideas as part of sqluff online: https://github.com/sqlfluff/sqlfluff-online/issues/14
- *The collection of issues related to macros and dbt libraries*:
  - https://github.com/sqlfluff/sqlfluff/issues/156
  - https://github.com/sqlfluff/sqlfluff/issues/371
  - https://github.com/sqlfluff/sqlfluff/issues/156
  - https://github.com/sqlfluff/sqlfluff/issues/381

I'm going to suggest a date of 24th October for closing the 0.4.0 iteration. By which time we either have to get clear on the second chunk above or we delay it to the 0.5.0 iteration. At that point I'll create the relevant issues and projects for 0.5.0.

**Comment by pwildenhain** [comment]
Reposted from: https://github.com/sqlfluff/sqlfluff/issues/444#issuecomment-703196892

I combed through the backlog today, trying to familiarize myself with the current state of the project, and organize for the upcoming release. Here's what I did:

### Add issues to the project shortlist

My criteria were:
- Issues that we in the old `0.4.0` milestone
- Bugs that seem like they have a fix in mind
- Features that I'm personally interested in

### Add new issue labels

Similar to how there were `parsing` and `new rule` labels, I wanted to add labels that reflected other parts of the package infrastructure.

I added:
- 🔗 [dialect](https://github.com/sqlfluff/sqlfluff/issues?q=is%3Aopen+is%3Aissue+label%3Adialect+). For issues related to specific SQL dialects. I toyed with the idea of a label for each dialect, but that seemed like overkill.
- 🔗 [templating](https://github.com/sqlfluff/sqlfluff/issues?q=is%3Aopen+is%3Aissue+label%3Atemplating). Mostly issues about `dbt` macros


### Thought about the 0.4.0 release

I think the major release should ideally squash as many bugs 🐛 👢 as possible. Which of these do y'all think we should tackle? 
🔗 [List of unassigned bugs](https://github.com/sqlfluff/sqlfluff/issues?q=is%3Aopen+is%3Aissue+label%3Abug+no%3Aassignee)

The other common thread that kept popping up was issues with how sqlfluff interacts with a`dbt` project. The ideas seemed to be numerous on how to best address this -- by far my favorite was to depend on `dbt` to handle `dbt` things (like macros). The thought being that if the user has templated SQL, they likely already have `dbt` installed for same project. Whatever the solution, I think this would also be a terrific ✨ enhancement for the next a major release.

and @barrywhart responded:
> 
> 
> We don't use dbt, but we do use templating. We might be in the minority here. What does "depending on dbt" look like? Is it as simple as creating a dbt configuration file, or something heavier? I think we can adapt if it's not too onerous.
> 
> 
> About bug fixes, 366 is an actual crash, so I'd say that's high priority. Ideally, I'd like to address all parsing issues, or if not, in priority order based on how many people will encounter them.

To which I said:
>I was thinking that dbt would more "enhance" the sqlfluff experience. Maybe a good solution looks like keeping all the jinja functionality we have now, and adding dbt specific support for macros and dbt packages. That you don't need dbt for sqlfluff templating to work unless you want tighter integration with your existing dbt project. 🤔 Don't know enough about the project though to suggest a more specific implementation

**Comment by barrywhart** [comment]
I would prefer that `sqlfluff` not require `dbt`. Given the number of subdependencies, it's a pretty heavyweight package. I just installed `dbt` into a new, empty virtualenv, and this is what I got:
```
agate==1.6.1
asn1crypto==1.4.0
attrs==20.2.0
azure-common==1.1.25
azure-core==1.8.2
azure-storage-blob==12.5.0
Babel==2.8.0
boto3==1.11.17
botocore==1.14.17
cachetools==4.1.1
certifi==2020.6.20
cffi==1.14.3
chardet==3.0.4
colorama==0.4.3
cryptography==2.9.2
dbt==0.18.0
dbt-bigquery==0.18.0
dbt-core==0.18.0
dbt-postgres==0.18.0
dbt-redshift==0.18.0
dbt-snowflake==0.18.0
decorator==4.4.2
docutils==0.15.2
google-api-core==1.16.0
google-auth==1.22.1
google-cloud-bigquery==1.25.0
google-cloud-core==1.3.0
google-resumable-media==0.5.1
googleapis-common-protos==1.6.0
hologram==0.0.10
idna==2.9
importlib-metadata==2.0.0
isodate==0.6.0
Jinja2==2.11.2
jmespath==0.10.0
json-rpc==1.13.0
jsonschema==3.1.1
leather==0.3.3
Logbook==1.5.3
MarkupSafe==1.1.1
minimal-snowplow-tracker==0.0.2
msrest==0.6.19
networkx==2.5
oauthlib==3.1.0
oscrypto==1.2.1
parsedatetime==2.6
protobuf==3.11.3
psycopg2-binary==2.8.6
pyasn1==0.4.8
pyasn1-modules==0.2.8
pycparser==2.20
pycryptodomex==3.9.8
PyJWT==1.7.1
pyOpenSSL==19.1.0
pyrsistent==0.17.3
python-dateutil==2.8.1
python-slugify==4.0.1
pytimeparse==1.1.8
pytz==2020.1
PyYAML==5.3.1
requests==2.23.0
requests-oauthlib==1.3.0
rsa==4.6
s3transfer==0.3.3
six==1.15.0
snowflake-connector-python==2.2.10
sqlparse==0.3.1
text-unidecode==1.3
typing-extensions==3.7.4.3
urllib3==1.25.10
vertical-prediction==0.1.0
Werkzeug==0.16.1
zipp==3.3.0
```

Whew!

Someone mentioned that `dbt` support could be implemented as a plugin. That is definitely possible. Another possibility that _may_ be simpler is to use `extras_require` (more info [here](https://stackoverflow.com/questions/41268863/difference-between-extras-require-and-install-requires-in-setup-py)). Essentially, `extras_require` allows delivering a single package with all the code, but in order to use certain features, you have to tell the package at installation time to install additional dependencies. For example, instead of `pip install sqlfluff`, you'd run `pip install sqlfluff[dbt]`.

Here's a real-world example of a package that uses this approach: [`apache-beam`](https://github.com/apache/beam/blob/master/sdks/python/setup.py#L308L316).

**Comment by pwildenhain** [comment]
Totally agree @barrywhart I think it should be an `extra_requires` or plugin.
@alanmcruickshank which of these ideas make more sense to you with the current package architecture?

**Comment by pwildenhain** [comment]
How are we feeling about our original deadline of tomorrow 😁 ?
Love that we got started on the public API with #501 I bet it'll make the online formatter and vscode plugin more robust. This seems like a good one to get a handle on before a major release.

I haven't seen/heard about any progress on the compatibility with `dbt` issues. This seems like a larger problem, that might need it's own release to focus on?

Overall, I'm really proud of everything that's been done since 0.3.0 and I think 0.4.0 is gonna be the new hotness 🔥

**Comment by alanmcruickshank** [comment]
Thanks for the reminder @pwildenhain !

I think given the level of clarity we've got on the dbt compatibility issues that for 0.4.0 we consider:
1. Implementing the plugin structure (I can't remember exactly which issue this was suggested in, but as a way to decouple anything we do for dbt from other parts of the codebase).
2. Using the plugin above (as one packaged automatically with sqlfluff for now), to manually support the majority of `dbt` and `dbt-utils` macros.
3. Documentation for power users to create their own plugins if they would like so that they can handle their own macros if they need to in future.
4. Some kind of source mapping as per #273 , unless it proves very unreliable, in which case we delay that.
5. [Potentially] pre-processing templated sql to better reflect whitespace linting as per the suggestion in #475 

I think the things we should push out of 0.4.0 based on the current level of clarity would be:
- letting dbt as a seperate package handle the rendering which we then deal with (because this entails a load of dependency issues and would require plugins *anyway*).
- actually splitting out parts of the current functionality and removing them from `sqlfluff`, requiring users who currently depend on them to install plugins (because we haven't demonstrated the plugins even work yet).

My reasoning for that split, is that the plugin architecture is the thing that introduces the basis for doing a load of the other things later. It's also a big enough architectural change that it deserves it's own major release. If in future we start splitting out parts of the codebase to be installed separately, then that would be something we do for 0.5.0.

How does that sound?

If we do go in that direction I think we should reword or recreate #371 , #381 & #156 to reflect the actual objectives we aim to chew off.

**Comment by pwildenhain** [comment]
🚀 Sounds good to me! I think plugin architectures are super cool 😎 and I can see how they'll level up ⬆️ this package

**Comment by alanmcruickshank** [comment]
I think I'm going to close the 0.4.0 release pipeline now. I've opened up a project for 0.5.0 to track any changes we want to put into that release. I'd like to aim to get the completed 0.4.0 out before Christmas now that most of the major pieces are at least in development.

I'll make a new issue for the major release schedule for 0.5.0 and we direct any discussion about future big changes there.

**Comment by pwildenhain** [comment]
Great, should we close this issue then?

**Comment by alanmcruickshank** [comment]
> Great, should we close this issue then?

Not yet, I think. I reckon we close it when we actually release 0.4.0 😄 .

@sqlfluff/sqlfluff-maintainers  It looks like everything is in progress. I had assumed that the plugin feature was originally planned for 0.4.0 but I can't see it in there. Do we want to attempt plugins for 0.4.0 or shall we get the current feature list out and then aim to target plugins as part of 0.5.0? It will be quite a big change so I think there are good reasons to go in either direction.

If not, then I'll do a sense check on the remaining items to makes sure we've got people working on each of them and then target a release _date_ 🏆 .

**Comment by pwildenhain** [comment]
I say push the plugins to `0.5.0` -- there are a bunch if improvements and bug fixes in here that I would like
- Users to try in the wild 🦁 
- Users to submit new bug reports on any new features we developed

**Comment by alanmcruickshank** [comment]
Heads up team: 0.4.0a1 is out for testing on pypi (https://pypi.org/project/sqlfluff/0.4.0a1/).

I'm sure I don't need to prompt people - but there are **a lot** of changes in here since 0.3.6, so I'd love to get some battle testing of this before we actually release it. If you do find issues, then please raise them in the usual way, but also make sure to mention that you're testing on 0.4.0a1 in the issue body.

The remaining items for development pre 0.4.0 are either small or already in progress so there may be additional features to land in the final deploy of 0.4.0 but I think we're close enough to start testing properly! Thanks to everyone for getting us this far. We're so close!

**Comment by pwildenhain** [comment]
Just tested the pre release on one of my project and it is BANGIN 💥 
⚡ Lightning fast fixing
🧠 Understands my `dbt` project

🙌 Great work y'all!

**Comment by barrywhart** [comment]
I'm testing 0.4.0a1.

**Comment by barrywhart** [comment]
I'm seeing an exception trace for one file:
```
CRITICAL   [L031] Applying rule L031 threw an Exception: 'NoneType' object has no attribute 'raw'
Traceback (most recent call last):
  File "/Users/bhart/.pyenv/versions/3.7.3/envs/sqlfluff-3.7.3/lib/python3.7/site-packages/sqlfluff/core/rules/base.py", line 289, in crawl
    dialect=dialect,
  File "/Users/bhart/.pyenv/versions/3.7.3/envs/sqlfluff-3.7.3/lib/python3.7/site-packages/sqlfluff/core/rules/std.py", line 3085, in _eval
    segment,
  File "/Users/bhart/.pyenv/versions/3.7.3/envs/sqlfluff-3.7.3/lib/python3.7/site-packages/sqlfluff/core/rules/std.py", line 3104, in _lint_aliases_in_join
    and base_table.raw == table_ref.raw
AttributeError: 'NoneType' object has no attribute 'raw'
```

**Comment by NiallRees** [comment]
> and base_table.raw == table_ref.raw

This one should be fixed by https://github.com/sqlfluff/sqlfluff/pull/581. Could you trying the latest master commit?

**Comment by barrywhart** [comment]
Overall, 0.4.0a1 worked well, just the one problem noted above. I ran `sqlfluff fix` on an entire project with about 5,000 lines of SQL. It took about 20 minutes the first time. The second run (i.e once the code was already clean) took 2 minutes.

I wonder if anyone has tried running SQLFluff using PyPy (much faster alternative to CPython). I'll see if I can get that working and report back.

**Comment by barrywhart** [comment]
I tested with PyPy in Docker on my Mac. It's a little faster, but not enough to really matter. 2:21 using CPython, 2:10 using PyPy. Typically I've seen much bigger speedups. At least it works!

**Comment by barrywhart** [comment]
I found a bug where a comment is deleted when `sqlfluff fix` reformats an `IF()`. I created an issue for it: https://github.com/sqlfluff/sqlfluff/issues/598.

**Comment by barrywhart** [comment]
@NiallRees: I just saw your note. I'll test this tomorrow.

**Comment by barrywhart** [comment]
@NiallRees: Confirmed, that issue is fixed now.

**Comment by pwildenhain** [comment]
Hey team 👋 

What do we think about re-grouping in January to discuss:
1. When to release 0.4
2. Plan out major features/timeline for 0.5?

**Comment by pwildenhain** [comment]
🔔 Any interest in a Feb meetup? I feel like 0.4.0 is in a good place, and there is a bunch of compelling stuff to start working on for 0.5.0

**Comment by NiallRees** [comment]
Yes @pwildenhain! I'll put a calendar invite for same day and time as last one next week!

**Comment by nolanbconaway** [comment]
Excellent idea! thx for the legwork @NiallRees

**Comment by alanmcruickshank** [comment]
Any last calls for 0.4.0? @sqlfluff/sqlfluff-maintainers 

I won't do the release today as planned (getting late my time), but I'll tidy up any remaining docs and a changelog during tomorrow. If there's anything still floating which we feel really needs to get in before the release then shout now 😄 .

I expect I'll release late morning UTC, assuming nothing else big comes up, but should have some time to handle anything and release later in the day if something comes up. 👍

**Comment by nolanbconaway** [comment]
I am so amped on this release!

**Comment by NiallRees** [comment]
> Any last calls for 0.4.0? @sqlfluff/sqlfluff-maintainers
> 
> I won't do the release today as planned (getting late my time), but I'll tidy up any remaining docs and a changelog during tomorrow. If there's anything still floating which we feel really needs to get in before the release then shout now 😄 .
> 
> I expect I'll release late morning UTC, assuming nothing else big comes up, but should have some time to handle anything and release later in the day if something comes up. 👍

This if poss! https://github.com/sqlfluff/sqlfluff/pull/603
